# **Online Retail Transaction Analysis**

## Objectives

* In this notebook, I will perform the ETL (Extract, Transform, Load) process on the Online Retail Transaction dataset. The aim is to clean, prepare, and transform the raw transaction data so that it is suitable for further analysis, visualisation, statistical testing, and predictive modelling.

* I will check the dataset for duplicate records, missing values, incorrect data types, negative or invalid values, and other inconsistencies. Necessary transformations will then be applied to improve data quality and create useful features for later stages of the project.

## Inputs

* The raw Online Retail Transaction dataset

* Python libraries

## Outputs

* A cleaned and transformed version of the dataset,ready for exploratory data analysis, hypothesis testing, visualisation, and machine learning.



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/Users/umarhussain/Documents/Online Retail Project/Online-Retail-Transaction-Analysis/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/umarhussain/Documents/Online Retail Project/Online-Retail-Transaction-Analysis'

# Extract, Transform, Load - ETL

Import Python libraries 

In [5]:
import pandas as pd
import numpy as np

Import raw dataset

In [6]:
# Load the raw dataset into a pandas DataFrame
df = pd.read_csv('Data/Online Retail.csv')

# Display the first five rows to check that the dataset loaded correctly
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


### Preview the Dataset
The first five and last five rows of the dataset are displayed to get an understanding of the data, including column names, values and overall structure of the dataset.

---

In [34]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


In [8]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680,France


The first and last five rows show that each row represents an individual transaction. This initial inspection helps confirm that the dataset has loaded correctly and provides an overview of the types of information available for further investigation and cleaning.

### Dataset Information
The dataset information is reviewed to understand its structure, column data types and whether any missing values are present before further investigation and cleaning.

In [9]:
# Display the dataset structure, column data types and non-null counts
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   541909 non-null  int64  
 7   Country      541909 non-null  str    
dtypes: float64(1), int64(2), str(5)
memory usage: 33.1 MB


The dataset contains 541,909 rows and 8 columns. The Description column contains 1,454 missing values, while the remaining columns contain no missing values. InvoiceDate is currently stored as a string and will need to be converted to datetime for time-based analysis.

### Missing Values
I will check the dataset for missing values in each column to identify any incomplete records that may require cleaning.

In [12]:
# Count the number of missing values in each column
df.isnull().sum()

InvoiceNo         0
StockCode         0
Description    1454
Quantity          0
InvoiceDate       0
UnitPrice         0
CustomerID        0
Country           0
dtype: int64

The missing-value check identified 1,454 missing values in the Description column. No missing values were identified in the other columns.

### Investigating Missing Product Descriptions

The Description column contains 1,454 missing values. I will examine these values to determine whether they contain useful transaction information.

In [16]:
# Display transactions where the product description is missing
missing_description = df[df['Description'].isna()]

missing_description.head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,15287,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,15287,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,15287,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,15287,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,15287,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,15287,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,15287,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,15287,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,15287,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,15287,United Kingdom


After the initial investigation the dataset shows that rows with missing product descriptions have a UnitPrice of 0.0 and include both positive and negative quantities.

In [22]:
# Count missing-description transactions with a UnitPrice of 0.0
(missing_description['UnitPrice'] == 0.0).sum()

np.int64(1454)

This shows all 1,454 rows with a missing product description also have a UnitPrice of 0.0. I will investigate quantity and customer patterns before deciding whether to remove them.

In [23]:
# Count negative quantities among rows with missing descriptions
(missing_description['Quantity'] < 0).sum()

np.int64(862)

In [24]:
# Count positive quantities among rows with missing descriptions
(missing_description['Quantity'] > 0).sum()

np.int64(592)

In [25]:
# Check how many unique customers appear in rows with missing descriptions
missing_description['CustomerID'].nunique()

1

I found that all 1,454 rows with missing product descriptions have a UnitPrice of 0. Of these rows, 862 have negative quantities and 592 have positive quantities. All rows are also associated with a single CustomerID. This pattern shows that these may not represent normal revenue-generating customer transactions.

In [26]:
# Display the most common invoice numbers among rows with missing descriptions
missing_description['InvoiceNo'].value_counts().head(20)

InvoiceNo
536414    1
536545    1
536546    1
536547    1
536549    1
536550    1
536552    1
536553    1
536554    1
536589    1
536764    1
536765    1
536995    1
536996    1
536997    1
536998    1
536999    1
537000    1
537001    1
537002    1
Name: count, dtype: int64

In [27]:
# Count the number of unique invoices among rows with missing descriptions
missing_description['InvoiceNo'].nunique()

1454

In [28]:
# Display the most common StockCodes among rows with missing descriptions
missing_description['StockCode'].value_counts().head(20)

StockCode
35965     10
23084     10
22084      9
22451      6
21784      5
22501      5
22866      5
23348      5
21067      5
21033      5
22139      4
84534B     4
22034      4
22114      4
21429      4
20966      4
22444      4
71477      4
21667      4
POST       4
Name: count, dtype: int64

In [29]:
# Investigate one affected StockCode across the full dataset
df[df['StockCode'] == '35965'][
    ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'UnitPrice']
].head(20)

,InvoiceNo,StockCode,Description,Quantity,UnitPrice
2889,536592,35965,FOLKART HEART NAPKIN RINGS,4,3.36
6017,536876,35965,FOLKART HEART NAPKIN RINGS,1,3.36
7205,537013,35965,NaN,-25,0.00
8071,537126,35965,FOLKART HEART NAPKIN RINGS,1,2.95
10678,537237,35965,FOLKART HEART NAPKIN RINGS,3,3.36
15508,537638,35965,FOLKART HEART NAPKIN RINGS,2,3.36
16653,537666,35965,FOLKART HEART NAPKIN RINGS,1,3.36
18410,537823,35965,FOLKART HEART NAPKIN RINGS,2,3.36
23007,538177,35965,FOLKART HEART NAPKIN RINGS,3,3.36
24762,538349,35965,FOLKART HEART NAPKIN RINGS,1,3.36


I investigated further and found that each row has a unique InvoiceNo and individual StockCodes showed that the same products appear elsewhere in the dataset with valid descriptions and positive selling prices. 

For example, StockCode 35965 appears as "FOLKART HEART NAPKIN RINGS" with positive UnitPrice values, while its missing-description row has a UnitPrice of 0 and a negative quantity.

This pattern shows that the missing-description rows are more likely to represent non-revenue activity rather than normal customer sales.

### Removing missing product descriptions

In [30]:
# Remove rows with missing descriptions identified as zero-value, non-standard transactions
df.dropna(subset=['Description'], inplace=True)

In [31]:
# Check the number of missing values in each column after cleaning
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [ ]:
# Check the number of rows and columns after removing missing product descriptions
df.shape

(540455, 8)

I have 1,454 rows with missing product descriptions, the dataset now contains 540,455 rows and 8 columns.

In [33]:
# Reset the index after removing rows
df.reset_index(drop=True, inplace=True)

### Check for duplicated rows
I will check for duplicate rows in the dataset. Duplicate transactions could lead to double counting and may distort sales and revenue.

In [35]:
# Count the number of duplicated rows in the dataset
df.duplicated().sum()

np.int64(5268)

In [39]:
# Calculate and round the percentage of duplicated rows
round((df.duplicated().sum() / len(df)) * 100, 2)

np.float64(0.97)

### Investigating duplicate rows
I checked for dupliclate rows and found 5,268 duplicated rows in the dataset, representing approximately 0.97% of the dataset. These rows will be inspected before removal to determine whether they represent duplicate transactions or legitimate repeated purchases.

In [36]:
# Display duplicated rows for investigation
df[df.duplicated(keep=False)].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920,United Kingdom


Inspection showed that these rows are identical across all columns, including InvoiceNo, StockCode, Quantity, InvoiceDate, UnitPrice and CustomerID.

### Remove duplicated rows

In [40]:
# Remove duplicate rows while keeping the first occurrence
df.drop_duplicates(inplace=True)

In [41]:
# Confirm no duplicate rows remain in the dataset
df.duplicated().sum()

np.int64(0)

In [42]:
# Check the dataset dimensions after removing duplicates
df.shape

(535187, 8)

In [43]:
# Reset the index after removing duplicate rows
df.reset_index(drop=True, inplace=True)

After removing 5,268 duplicate rows, the dataset contains 535,187 rows and 8 columns.

### Investigating Quantity and UnitPrice

The Quantity and UnitPrice columns will be examined for zero, negative and unusually large values. Negative quantities may represent returned or cancelled transactions.

In [44]:
# Display summary statistics for Quantity and UnitPrice
df[['Quantity', 'UnitPrice']].describe()

,Quantity,UnitPrice
count,535187.000000,535187.000000
mean,9.671593,4.645242
std,219.059056,97.364810
min,-80995.000000,-11062.060000
25%,1.000000,1.250000
50%,3.000000,2.080000
75%,10.000000,4.130000
max,80995.000000,38970.000000


The summary statistics show that Quantity ranges from -80,995 to 80,995, while UnitPrice ranges from -11,062.06 to 38,970.

In [45]:
# Count rows with a negative Quantity
(df['Quantity'] < 0).sum()

np.int64(9725)

In [46]:
# Display rows with negative quantities
df[df['Quantity'] < 0].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548,United Kingdom
915,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897,United Kingdom


In [47]:
# Count rows with a negative UnitPrice
(df['UnitPrice'] < 0).sum()

np.int64(2)

In [48]:
# Display rows with negative UnitPrice values
df[df['UnitPrice'] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
296537,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,15287,United Kingdom
296538,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,15287,United Kingdom


In [49]:
# Count rows where UnitPrice is zero
(df['UnitPrice'] == 0).sum()

np.int64(1056)

In [50]:
# Display rows where UnitPrice is zero
df[df['UnitPrice'] == 0].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6282,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,15287,United Kingdom
6283,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,15287,United Kingdom
7169,537032,21275,?,-30,2010-12-03 16:50:00,0.0,15287,United Kingdom
9090,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647,Germany
12917,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,15287,United Kingdom
12918,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,15287,United Kingdom
12962,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,15287,United Kingdom
14020,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,15287,United Kingdom
14021,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,15287,United Kingdom
14022,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,15287,United Kingdom


In [51]:
# Display the row with the largest Quantity
df.loc[df['Quantity'].idxmax()]

InvoiceNo                           581483
StockCode                            23843
Description    PAPER CRAFT , LITTLE BIRDIE
Quantity                             80995
InvoiceDate            2011-12-09 09:15:00
UnitPrice                             2.08
CustomerID                           16446
Country                     United Kingdom
Name: 533706, dtype: object

In [52]:
# Display the row with the smallest Quantity
df.loc[df['Quantity'].idxmin()]

InvoiceNo                          C581484
StockCode                            23843
Description    PAPER CRAFT , LITTLE BIRDIE
Quantity                            -80995
InvoiceDate            2011-12-09 09:27:00
UnitPrice                             2.08
CustomerID                           16446
Country                     United Kingdom
Name: 533707, dtype: object

In [53]:
# Display the row with the highest UnitPrice
df.loc[df['UnitPrice'].idxmax()]

InvoiceNo                  C556445
StockCode                        M
Description                 Manual
Quantity                        -1
InvoiceDate    2011-06-10 15:31:00
UnitPrice                  38970.0
CustomerID                   15098
Country             United Kingdom
Name: 219961, dtype: object

In [54]:
# Display the row with the lowest UnitPrice
df.loc[df['UnitPrice'].idxmin()]

InvoiceNo                  A563186
StockCode                        B
Description        Adjust bad debt
Quantity                         1
InvoiceDate    2011-08-12 14:51:00
UnitPrice                -11062.06
CustomerID                   15287
Country             United Kingdom
Name: 296537, dtype: object

The dataset contains 9,725 rows with negative quantities. Inspection showed that many of these rows have invoice numbers beginning with "C", indicating cancelled transactions. 

The largest positive quantity of 80,995 was followed shortly afterwards by a corresponding quantity of -80,995 for the same product and customer,  supporting that negative quantities represent cancellations or returns. These rows will be retained.

In [55]:
# Count negative-quantity rows with cancellation invoice numbers
df[df['Quantity'] < 0]['InvoiceNo'].str.startswith('C').sum()

np.int64(9251)

In [59]:
# Display negative-quantity rows that do not have a cancellation invoice number
negative_not_cancelled = df[
    (df['Quantity'] < 0) &
    (~df['InvoiceNo'].str.startswith('C'))
]

negative_not_cancelled.head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
7169,537032,21275,?,-30,2010-12-03 16:50:00,0.0,15287,United Kingdom
12917,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,15287,United Kingdom
12918,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,15287,United Kingdom
12962,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,15287,United Kingdom
20935,538072,22423,faulty,-13,2010-12-09 14:10:00,0.0,15287,United Kingdom
21114,538090,20956,?,-723,2010-12-09 14:48:00,0.0,15287,United Kingdom
21870,538161,46000S,Dotcom sales,-100,2010-12-09 17:25:00,0.0,15287,United Kingdom
21871,538162,46000M,Dotcom sales,-100,2010-12-09 17:25:00,0.0,15287,United Kingdom
41929,540010,22501,reverse 21/5/10 adjustment,-100,2011-01-04 11:13:00,0.0,15287,United Kingdom
41930,540012,22502,reverse 21/5/10 adjustment,-100,2011-01-04 11:14:00,0.0,15287,United Kingdom


In [60]:
len(negative_not_cancelled)

474

In [61]:
# Display the most common descriptions among negative non-cancellation rows
negative_not_cancelled['Description'].value_counts().head(20)

Description
check                         120
damages                        45
damaged                        42
?                              41
sold as set on dotcom          20
Damaged                        14
thrown away                     9
Unsaleable, destroyed.          9
??                              7
damages?                        5
wet damaged                     5
ebay                            5
smashed                         4
missing                         3
CHECK                           3
wet pallet                      3
Dotcom sales                    2
reverse 21/5/10 adjustment      2
counted                         2
crushed                         2
Name: count, dtype: int64

In [62]:
# Display the most common StockCodes among negative non-cancellation rows
negative_not_cancelled['StockCode'].value_counts().head(20)

StockCode
85175     5
85172     4
21830     4
72802C    4
22423     3
46000S    3
22719     3
21135     3
72802B    3
23343     3
72807A    3
72807B    3
72807C    3
20956     2
22501     2
84406B    2
21621     2
22797     2
84249A    2
20893     2
Name: count, dtype: int64

In [63]:
# Check whether all negative non-cancellation rows have a UnitPrice of zero
(negative_not_cancelled['UnitPrice'] == 0).sum()

np.int64(474)

In [64]:
# Remove negative stock-adjustment rows that are not customer cancellations
df.drop(
    df[
        (df['Quantity'] < 0) &
        (~df['InvoiceNo'].str.startswith('C'))
    ].index,
    inplace=True
)

In [65]:
df[
    (df['Quantity'] < 0) &
    (~df['InvoiceNo'].str.startswith('C'))
].shape[0]

0

In [66]:
df.shape

(534713, 8)

All 474 negative-quantity rows without a cancellation invoice number also had a UnitPrice of 0. Their descriptions included terms such as "damages", "check", "missing" and "thrown away", showing that they represent internal stock adjustments rather than customer transactions. These rows were removed from the dataset.

### Removing Negative UnitPrice Values

Two rows were identified with negative UnitPrice values. Both rows were labelled "Adjust bad debt", showing that they represent accounting adjustments rather than customer retail transactions. As this project focuses on sales and customer spending, these rows will be removed from the dataset.

In [67]:
# Remove rows with negative UnitPrice values identified as bad debt adjustments
df.drop(
    df[df['UnitPrice'] < 0].index,
    inplace=True
)

In [68]:
# Confirm that no negative UnitPrice values remain
(df['UnitPrice'] < 0).sum()

np.int64(0)

In [69]:
df.shape

(534711, 8)

The two negative UnitPrice rows identified as bad debt adjustments were removed. The dataset now contains 534,711 rows and 8 columns, with no remaining negative UnitPrice values.

### Investigate Zero UnitPrice Values
Rows with a UnitPrice of 0 will be investigated to determine whether they represent genuine customer transactions, free items, or non-sales activity. These rows will be examined before deciding whether they should be retained or removed.

In [70]:
# Count rows where UnitPrice is zero
(df['UnitPrice'] == 0).sum()

np.int64(582)

In [71]:
# Display rows where UnitPrice is zero
df[df['UnitPrice'] == 0].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6282,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,15287,United Kingdom
6283,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,15287,United Kingdom
9090,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647,Germany
14020,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,15287,United Kingdom
14021,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,15287,United Kingdom
14022,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,15287,United Kingdom
14023,537534,48184,DOORMAT ENGLISH ROSE,3,2010-12-07 11:48:00,0.0,15287,United Kingdom
14024,537534,48111,DOORMAT 3 SMILEY CATS,1,2010-12-07 11:48:00,0.0,15287,United Kingdom
14025,537534,22697,GREEN REGENCY TEACUP AND SAUCER,1,2010-12-07 11:48:00,0.0,15287,United Kingdom
14026,537534,22682,FRENCH BLUE METAL DOOR SIGN 7,1,2010-12-07 11:48:00,0.0,15287,United Kingdom


In [72]:
# Display the most common descriptions among zero-price rows
df[df['UnitPrice'] == 0]['Description'].value_counts().head(20)

Description
check                              39
found                              25
adjustment                         14
FRENCH BLUE METAL DOOR SIGN 1       9
amazon                              8
FRENCH BLUE METAL DOOR SIGN 8       8
Found                               8
FRENCH BLUE METAL DOOR SIGN 4       7
FRENCH BLUE METAL DOOR SIGN No      7
OWL DOORSTOP                        7
FRENCH BLUE METAL DOOR SIGN 3       7
RECIPE BOX PANTRY YELLOW DESIGN     7
Amazon                              7
FRENCH BLUE METAL DOOR SIGN 7       6
FRENCH BLUE METAL DOOR SIGN 5       6
FRENCH BLUE METAL DOOR SIGN 6       6
RED KITCHEN SCALES                  6
?                                   6
Manual                              6
BOX OF 24 COCKTAIL PARASOLS         5
Name: count, dtype: int64

In [73]:
# Count zero-price rows with positive quantities
((df['UnitPrice'] == 0) & (df['Quantity'] > 0)).sum()

np.int64(582)

In [74]:
# Count zero-price rows with invoice numbers beginning with C
df[df['UnitPrice'] == 0]['InvoiceNo'].str.startswith('C').sum()

np.int64(0)

In [75]:
# Display the most common CustomerIDs among zero-price rows
df[df['UnitPrice'] == 0]['CustomerID'].value_counts().head(10)

CustomerID
15287    542
13081      4
14646      4
14911      2
12415      2
13985      2
12647      1
16560      1
15107      1
17560      1
Name: count, dtype: int64

In [80]:
# Remove rows with a UnitPrice of zero as they do not represent revenue-generating sales
df.drop(
    df[df['UnitPrice'] == 0].index,
    inplace=True
)

In [81]:
# Confirm that no zero UnitPrice rows remain
(df['UnitPrice'] == 0).sum()

np.int64(0)

In [82]:
# Check the dataset dimensions after removing zero-price rows
df.shape

(534129, 8)

All 582 rows had positive quantities, while none had negative quantities. Investigating the descriptions identified terms such as "check", "found", "adjustment", "amazon" and "Manual", alongside some normal product descriptions recorded at no charge.

As these rows do not represent revenue-generating transactions the rows will be removed from the  dataset to prevent them from distorting transaction and quantity-based analysis.

### Investigating Cancellations and Returns

Negative quantities may represent returned or cancelled customer transactions. Invoice numbers beginning with "C" will be investigated to determine whether they identify as cancellations and whether these transactions should be retained for returns and net revenue analysis.

In [85]:
# Count rows with invoice numbers beginning with 'C'
df['InvoiceNo'].str.startswith('C').sum()

np.int64(9251)

In [87]:
# Count unique cancellation invoice numbers
df[df['InvoiceNo'].str.startswith('C')]['InvoiceNo'].nunique()

3836

In [88]:
# Display examples of cancelled transactions
df[df['InvoiceNo'].str.startswith('C')].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548,United Kingdom
915,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897,United Kingdom


In [89]:
# Count cancellation rows with negative quantities
(
    (df['InvoiceNo'].str.startswith('C')) &
    (df['Quantity'] < 0)
).sum()

np.int64(9251)

In [90]:
# Check for cancellation invoices without negative quantities
df[
    (df['InvoiceNo'].str.startswith('C')) &
    (df['Quantity'] >= 0)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


In [91]:
# Count remaining rows with negative quantities
(df['Quantity'] < 0).sum()

np.int64(9251)

In [92]:
# Check for negative quantities without a cancellation invoice number
df[
    (df['Quantity'] < 0) &
    (~df['InvoiceNo'].str.startswith('C'))
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


In [93]:
# Calculate the percentage of rows representing cancellations
round(
    df['InvoiceNo'].str.startswith('C').sum() / len(df) * 100,
    2
)

np.float64(1.73)

In [94]:
# Display the most frequently cancelled products
df[df['InvoiceNo'].str.startswith('C')]['Description'].value_counts().head(10)

Description
Manual                               244
REGENCY CAKESTAND 3 TIER             180
POSTAGE                              126
JAM MAKING SET WITH JARS              87
Discount                              77
SET OF 3 CAKE TINS PANTRY DESIGN      73
SAMPLES                               60
ROSES REGENCY TEACUP AND SAUCER       54
STRAWBERRY CERAMIC TRINKET BOX        54
RECIPE BOX PANTRY YELLOW DESIGN       47
Name: count, dtype: int64

The investigation identified 9,251 rows with invoice numbers beginning with "C", representing 3,836 unique cancellation invoices. All of these rows have negative quantities. 

Cancellation and return rows represent approximately 1.73% of the cleaned dataset. These rows will be retained because they contain useful information for analysing cancellations, returns and their impact on net revenue.

### Creating Transaction Type

A new TransactionType column will be created to distinguish normal sales from cancelled or returned transactions.

In [97]:
# Classify each transaction as a cancellation or regular sale
df["TransactionType"] = np.where(
    df["InvoiceNo"].str.startswith("C"),
    "Cancellation/Return",
    "Sale"
)

In [98]:
# Check the number of transactions in each category
df["TransactionType"].value_counts()

TransactionType
Sale                   524878
Cancellation/Return      9251
Name: count, dtype: int64

In [99]:
# Preview the new TransactionType column
df[['InvoiceNo', 'Quantity', 'TransactionType']].head(20)

,InvoiceNo,Quantity,TransactionType
0,536365,6,Sale
1,536365,6,Sale
2,536365,8,Sale
3,536365,6,Sale
4,536365,6,Sale
5,536365,2,Sale
6,536365,6,Sale
7,536366,6,Sale
8,536366,6,Sale
9,536367,32,Sale


### Review Column Names

The column names will be reviewed to ensure that they are clear and easy to understand.

In [101]:
# Display the current column names
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country', 'TransactionType'],
      dtype='str')

The existing column names are clear and are easy to understand. Therefore, no column names need to be changed.

### Review Data Types

The current data types will be reviewed to identify any columns that require conversion before analysis. 

In [102]:
# Review the current data types
df.dtypes

InvoiceNo              str
StockCode              str
Description            str
Quantity             int64
InvoiceDate            str
UnitPrice          float64
CustomerID           int64
Country                str
TransactionType        str
dtype: object

### Convert InvoiceDate to Datetime

InvoiceDate will be converted to datetime format. This will allow time-based analysis to be performed, including monthly sales trends and customer purchasing behaviour over time.

In [103]:
# Convert InvoiceDate to datetime format
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [104]:
# Confirm that InvoiceDate has been converted successfully
df['InvoiceDate'].dtype

dtype('<M8[us]')

InvoiceDate was successfully converted to datetime format, making the column suitable for time-based analysis.

### Convert Identifier Data Types

CustomerID represents a customer identifier rather than a numerical measurement. Therefore, it will be converted to a string data type. InvoiceNo and StockCode will also be treated as identifiers to ensure that they are not interpreted as continuous numerical variables.

In [105]:
# Convert identifier columns to string data types
df['CustomerID'] = df['CustomerID'].astype(str)
df['InvoiceNo'] = df['InvoiceNo'].astype(str)
df['StockCode'] = df['StockCode'].astype(str)

In [106]:
# Check the data types after conversion
df.dtypes

InvoiceNo                     str
StockCode                     str
Description                   str
Quantity                    int64
InvoiceDate        datetime64[us]
UnitPrice                 float64
CustomerID                    str
Country                       str
TransactionType               str
dtype: object

The identifier columns are now stored as text rather than being treated as numerical measurements. Quantity and UnitPrice remain numerical because they are required for calculations.

### Create Total Transaction Column

A new TotalTransactionValue column will be created by multiplying Quantity by UnitPrice.

Positive values represent regular sales, while negative values represent cancellations or returns.

In [119]:
# Calculate transaction value and round to two decimal places
df['TotalTransactionValue'] = (
    df['Quantity'] * df['UnitPrice']
).round(2)

In [120]:
# Preview the new transaction value column
df[['Quantity', 'UnitPrice', 'TransactionType', 'TotalTransactionValue']].head(20)

,Quantity,UnitPrice,TransactionType,TotalTransactionValue
0,6,2.55,Sale,15.30
1,6,3.39,Sale,20.34
2,8,2.75,Sale,22.00
3,6,3.39,Sale,20.34
4,6,3.39,Sale,20.34
5,2,7.65,Sale,15.30
6,6,4.25,Sale,25.50
7,6,1.85,Sale,11.10
8,6,1.85,Sale,11.10
9,32,1.69,Sale,54.08


In [121]:
# Confirm that cancellations produce negative transaction values
df[df['TransactionType'] == 'Cancellation/Return'][
    ['Quantity', 'UnitPrice', 'TransactionType', 'TotalTransactionValue']
].head(20)

,Quantity,UnitPrice,TransactionType,TotalTransactionValue
141,-1,27.50,Cancellation/Return,-27.50
154,-1,4.65,Cancellation/Return,-4.65
235,-12,1.65,Cancellation/Return,-19.80
236,-24,0.29,Cancellation/Return,-6.96
237,-24,0.29,Cancellation/Return,-6.96
238,-24,0.29,Cancellation/Return,-6.96
239,-12,3.45,Cancellation/Return,-41.40
240,-12,1.65,Cancellation/Return,-19.80
241,-24,1.65,Cancellation/Return,-39.60
915,-6,4.25,Cancellation/Return,-25.50


TotalTransactionValue was calculated successfully. Regular sales produce positive transaction values, while cancellations produce negative values. 

### Create Date Columns

Additional date columns will be extracted from InvoiceDate for time-based analysis. Year, Month and YearMonth will be created so that sales and customer spending can be analysed over different time periods.

In [122]:
# Extract the transaction year from InvoiceDate
df['Year'] = df['InvoiceDate'].dt.year

In [123]:
# Extract the transaction month from InvoiceDate
df['Month'] = df['InvoiceDate'].dt.month

In [124]:
# Create a YearMonth variable for monthly trend analysis
df['YearMonth'] = df['InvoiceDate'].dt.to_period('M').astype(str)

In [126]:
# Preview the newly created date features
df[
    ['InvoiceDate', 'Year', 'Month', 'YearMonth']
].head()

,InvoiceDate,Year,Month,YearMonth
0,2010-12-01 08:26:00,2010,12,2010-12
1,2010-12-01 08:26:00,2010,12,2010-12
2,2010-12-01 08:26:00,2010,12,2010-12
3,2010-12-01 08:26:00,2010,12,2010-12
4,2010-12-01 08:26:00,2010,12,2010-12


Year, Month and YearMonth were successfully created from InvoiceDate. These features will support later analysis of yearly and monthly sales trends, with YearMonth providing a chronological monthly variable across both 2010 and 2011.

### Preview Transformed Dataset

The transformed dataset will be previewed to confirm that the newly created variables have been added correctly.

In [127]:
# Preview the transformed dataset
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TransactionType,TotalTransactionValue,Year,Month,YearMonth
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,Sale,15.30,2010,12,2010-12
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,Sale,20.34,2010,12,2010-12
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,Sale,22.00,2010,12,2010-12
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,Sale,20.34,2010,12,2010-12
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,Sale,20.34,2010,12,2010-12


## Data Cleaning Summary

The raw dataset initially contained 541,909 rows and 8 columns. During the ETL process, the dataset was investigated and cleaned.

A total of 1,454 rows with missing product descriptions were investigated and removed after they were identified as zero-value, non-standard transactions. Exact duplicate rows were also identified and removed to prevent double-counting.

I investigated Quantity and UnitPrice for invalid values. Negative quantities associated with cancellation invoices were retained, while 474 negative-quantity stock adjustment rows were removed. Two negative UnitPrice rows relating to bad debt adjustments and 582 zero-price non-revenue rows were also removed.

Transactions that represented cancellations were identified using invoice numbers beginning with "C" and retained because they are required for returns and net revenue analysis.

InvoiceDate was converted to datetime format and identifier columns were assigned appropriate data types. Additional features were then created for TransactionType, TotalTransactionValue, Year, Month and YearMonth.

## Load Cleaned Dataset

Following the transformation stage, the cleaned dataset will be exported as a CSV file. This cleaned dataset will be used for the rest of the project.

In [143]:
# Export the cleaned dataset for use in later analysis
df.to_csv('Data/Cleaned_Online_Retail_Dataset.csv', index=False)